src/back/spv/image.rs

```diff
@@ -28,6 +28,111 @@ struct TextureCoordinates {
     size: Option<crate::VectorSize>,
 }
 
+/// Information about image access that is indifferent to bounds checks.
+///
+/// When generating code for `ImageLoad` and `ImageStore` expressions, the image
+/// bounds checks policy can affect some operands of the image access
+/// instruction (the coordinates, level of detail, and sample index), but other
+/// aspects are unaffected: the image id, result type, and the specific SPIR-V
+/// instruction used.
+///
+/// This struct holds the latter category of information, saving us from passing
+/// a half-dozen parameters along the various code paths. The parts that are
+/// affected by bounds checks, are passed as parameters to the `write`
+/// method.
+struct ImageAccess {
+    /// The specific opcode we'll use to perform the fetch. Storage textures
+    /// require `OpImageRead`, while other textures require `OpImageFetch`.
+    opcode: spirv::Op,
+
+    /// The type id produced by the actual image access instruction.
+    type_id: Word,
+
+    /// The id of the image being accessed.
+    image_id: Word,
+}
+
+impl ImageAccess {
+    fn from_image_expr(
+        ctx: &mut BlockContext<'_>,
+        image_id: Word,
+        image_class: crate::ImageClass,
+        result_type_id: Word,
+    ) -> Result<ImageAccess, Error> {
+        let opcode = match image_class {
+            crate::ImageClass::Storage { .. } => spirv::Op::ImageRead,
+            crate::ImageClass::Depth { .. } | crate::ImageClass::Sampled { .. } => {
+                spirv::Op::ImageFetch
+            }
+        };
+
+        // `OpImageRead` and `OpImageFetch` instructions produce vec4<f32>
+        // values. Most of the time, we can just use `result_type_id` for
+        // this. The exception is that `Expression::ImageLoad` from a depth
+        // texture produces a scalar `f32`, so in that case we need to find
+        // the right SPIR-V type for the access instruction here.
+        let type_id = match image_class {
+            crate::ImageClass::Depth { .. } => {
+                ctx.get_type_id(LookupType::Local(LocalType::Value {
+                    vector_size: Some(crate::VectorSize::Quad),
+                    kind: crate::ScalarKind::Float,
+                    width: 4,
+                    pointer_class: None,
+                }))
+            }
+            _ => result_type_id,
+        };
+
+        Ok(ImageAccess {
+            opcode,
+            type_id,
+            image_id,
+        })
+    }
+
+    /// Write an instruction to access a given texel of this image.
+    ///
+    /// This assumes that `coordinates_id` has already had the image array index,
+    /// if any, folded in, as done by `write_texture_coordinates`.
+    ///
+    /// Return the value id produced by the instruction.
+    fn write(
```

Comment:
alternatively, this could just return `Instruction` and avoid receiving `id_gen` and `block`

Reply:
This ends up needing to work this way in the versions that support stores.